pkg/providers/instancetype/instancetype.go

```diff
@@ -93,13 +97,126 @@ func (p *DefaultProvider) List(ctx context.Context, kc *v1alpha1.KubeletConfigur
+	}
+
+	if p.cm.HasChanged("instance-types", instanceTypes) {
+		// Only update instanceTypesSeqNun with the instance types have been changed
+		// This is to not create new keys with duplicate instance types option
+		atomic.AddUint64(&p.instanceTypesSeqNum, 1)
+		log.FromContext(ctx).WithValues(
+			"count", len(instanceTypes)).V(1).Info("discovered instance types")
+	}
+	p.instanceTypesInfo = instanceTypes
 
-	// TODO: implement me
 	return nil
 }
 
 func (p *DefaultProvider) UpdateInstanceTypeOfferings(ctx context.Context) error {
+	// DO NOT REMOVE THIS LOCK ----------------------------------------------------------------------------
+	// We lock here so that multiple callers to getInstanceTypeOfferings do not result in cache misses and multiple
+	// calls to EC2 when we could have just made one call.
+
+	p.muInstanceTypeOfferings.Lock()
+	defer p.muInstanceTypeOfferings.Unlock()
+
+	// Get offerings from ECS
+	instanceTypeOfferings := map[string]sets.Set[string]{}
+	describeAvailableResourceRequest := &ecsclient.DescribeAvailableResourceRequest{
+		RegionId:            tea.String(p.region),
+		DestinationResource: tea.String("InstanceType"),
+	}
+	runtime := &util.RuntimeOptions{}
 
-	// TODO: implement me
+	// TODO: we may use other better API in the future.
+	resp, err := p.ecsClient.DescribeAvailableResourceWithOptions(describeAvailableResourceRequest, runtime)
+	if err != nil {
+		return err
+	}
+
+	if resp.Body == nil || resp.Body.AvailableZones == nil || len(resp.Body.AvailableZones.AvailableZone) == 0 {
+		return errors.New("DescribeAvailableResourceWithOptions failed to return any instance types")
+	}
+
+	for _, az := range resp.Body.AvailableZones.AvailableZone {
+		if *az.StatusCategory == "WithStock" { // WithStock, ClosedWithStock, WithoutStock, ClosedWithoutStock
+			processAvailableResources(az, instanceTypeOfferings)
+		}
+	}
+
+	if p.cm.HasChanged("instance-type-offering", instanceTypeOfferings) {
+		// Only update instanceTypesSeqNun with the instance type offerings  have been changed
+		// This is to not create new keys with duplicate instance type offerings option
+		atomic.AddUint64(&p.instanceTypeOfferingsSeqNum, 1)
+		log.FromContext(ctx).WithValues("instance-type-count", len(instanceTypeOfferings)).V(1).Info("discovered offerings for instance types")
+	}
+	p.instanceTypeOfferings = instanceTypeOfferings
 	return nil
 }
+
+func processAvailableResources(az *ecsclient.DescribeAvailableResourceResponseBodyAvailableZonesAvailableZone, instanceTypeOfferings map[string]sets.Set[string]) {
+	if az.AvailableResources == nil || az.AvailableResources.AvailableResource == nil {
+		return
+	}
+
+	for _, ar := range az.AvailableResources.AvailableResource {
+		if ar.SupportedResources == nil || ar.SupportedResources.SupportedResource == nil {
+			continue
+		}
+
+		for _, sr := range ar.SupportedResources.SupportedResource {
+			if *sr.StatusCategory == "WithStock" { // WithStock, ClosedWithStock, WithoutStock, ClosedWithoutStock
```

Comment:
I think we should also consider ClosedWithStock.

Reply:
+1

Reply:
I think I'll use WithStock first, and then consider whether to use it after I've tested the margins of the ClosedWithStock state.

Reply:
Can you help add a TODO comment? we will back to this latter